Add subreddit creation and update_settings

voussoir opened the pull request:
### What's new

- Add class `SubredditHelper`.
- Remove method `Reddit.subreddit`.
- Add staticmethod `Subreddit._create_or_update` to access the site_admin endpoint.
- Add method `SubredditHelper.create`. The required fields are `name`, `title`, `link_type`, `subreddit_type`, `wikimode`. Everything else can be left for reddit to decide.
- Add method `SubredditModeration.settings` to fetch the current settings.
- Add method `SubredditModeration.update_settings` to overwrite specific settings. In my opinion, `set_settings` from PRAW3 should not be re-implemented, updating is where it's at.
- Add test `TestSubreddit.test_create`
- Add test `TestSubredditModeration.test_update_settings`.
- Move multireddit creation test from `test_reddit.py` to `test_multi.py`.
- Rename mispelled cassettes "Mu**li**reddit" -> "Mu**lti**reddit".

### Concerns

- There's a subreddit setting called "domain", and I can't find any documentation for what it does. It's not listed on the [site docs](https://www.reddit.com/dev/api/oauth#POST_api_site_admin) and I don't know what [this](https://github.com/reddit/reddit/blob/4b2a669d6fe85f85c518a06a905ef1c0cfa57478/r2/r2/controllers/api.py#L2854-L2857) or [this](https://github.com/reddit/reddit/blob/4b2a669d6fe85f85c518a06a905ef1c0cfa57478/r2/r2/models/subreddit.py#L494-L510) mean. It's included in the `_create_or_update` parameters just for completeness sake.

&nbsp;

It's been a while. Please critique!

bboe commented on praw/models/helpers.py:
```diff
@@ -66,10 +66,44 @@ def create(self, display_name, subreddits, description_md=None,
+        """
+        Create a new subreddit.
+
+        See :meth:`~Subreddit._create_or_update` for details on each parameter.
```
Can we actually put all the docstring parameters here? Since `_create_or_update` is not part of the public interface I would prefer not referencing its docstring.

That means, yes, we need to duplicate the parameters in the update function too (though I suppose you could reference that from here.

bboe commented on praw/models/helpers.py:
```diff
@@ -66,10 +66,44 @@ def create(self, display_name, subreddits, description_md=None,
+        return Subreddit(self._reddit, display_name=display_name)
+
+    def create(self, name, link_type, subreddit_type, title, wikimode,
+               **kwargs):
```
Can we use a more descriptive name than kwargs? I know that's a python convention, but maybe something like `other_settings` provides more description to what we're trying to capture with the `**`.

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+                          title=None, wiki_edit_age=None, wiki_edit_karma=None,
+                          wikimode=None):
+        """Create a new subreddit.
+
```
With moving the parameter explanation to other functions, I don't think this docstring is necessary on this function.

bboe commented:
> There's a subreddit setting called "domain", and I can't find any documentation for what it does.

The domain feature is supposed to allow you to map a domain name like reddit.bryceboe.com to a subreddit, e.g., /r/bboe. I could configure DNS for reddit.bryceboe.com with a CNAME that points to bboe.reddit.com, and update the domain settings for the subreddit to add `reddit.bryceboe.com`.

Here's two discussions regarding the setting, which may not work very well these days:
https://www.reddit.com/r/modhelp/comments/eju7m/what_is_the_point_of_the_domaincname_config_option/
https://www.reddit.com/r/help/comments/cguj2/trying_to_use_a_custom_domain_for_my_subreddit/

For the setting you could write, "domain name with a name that points to SUBREDDIT.reddit.com."

> In my opinion, set_settings from PRAW3 should not be re-implemented, updating is where it's at.

I agree. If someone asks for it in the future, they can add it. :)

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+        :param link_type: The types of submissions users can make.
+            One of ``any``, ``link``, ``self``.
+        :param name: The name for the new subreddit. Required during creation.
+        :param over_18: Viewers must be over 18 years old.
```
Can we put, "i.e., NSFW" since that's what people commonly refer to this setting as?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+        :param sr: The fullname of the subreddit whose settings will be
+            updated.
+        :param submit_link_label: Custom label for submit link button
+            (blank for default).
```
Is `''` for default, or `None`?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+            (blank for default).
+        :param submit_text: Text to show on submission page.
+        :param submit_text_label: Custom label for submit text post button
+            (blank for default).
```
Same question, it's not clear what blank means (if this is used elsewhere then it probably should be made more clear there as well.

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+        :param submit_text: Text to show on submission page.
+        :param submit_text_label: Custom label for submit text post button
+            (blank for default).
+        :param subreddit_type: One of ``gold_restricted``, ``archived``,
```
Can you sort these options?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+        :param public_traffic: Make the traffic stats page public.
+        :param show_media: Expand media previews on comments pages.
+        :param spam_comments: Spam filter strength for comments.
+            One of ``low``, ``high``, ``all``.
```
Can you sort these options?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+            ``public``.
+        :param suggested_comment_sort: All comment threads will use this
+            sorting method by default. Leave blank, or choose one of
+            ``confidence``, ``top``, ``new``, ``controversial``, ``old``,
```
Can you sort these options?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+            sorting method by default. Leave blank, or choose one of
+            ``confidence``, ``top``, ``new``, ``controversial``, ``old``,
+            ``random``, ``qa``.
+        :param title: The title for the new subreddit.
```
The title can be changed, right? So maybe omit "new".

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+            ``confidence``, ``top``, ``new``, ``controversial``, ``old``,
+            ``random``, ``qa``.
+        :param title: The title for the new subreddit.
+        :param wiki_edit_age: Account age (days) required to edit and create
```
Can we say "Account age, in days, ..."?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+            wiki pages.
+        :param wikimode: One of ``disabled``, ``modonly``, ``anyone``.
+
+        Any parameters left `None` will take on a default value assigned by
```
Can we say, "Any keyword parameters not provided, or explicitly set to None, will ... assigned by the Reddit server"?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
+        Any parameters left `None` will take on a default value assigned by
+        the server.
+
+        :returns: The server response body.
```
Do we need this response? I've been questioning the useless responses from Reddit. PRAW does a pretty good job of raising an exception when something fails. If the success response provides no useful information, then I think we should not have a return (I likely need to fix this elsewhere). We can always add one later if it turns out we were wrong, or reddit starts returning useful information.

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -271,6 +382,11 @@ def remove(self, thing, spam=False):
         self.subreddit._reddit.post(API_PATH['remove'], data=data)
 
+    def settings(self):
+        """Get the subreddit's current settings."""
```
Can this say, "Return a dictionary of the subreddit's current settings."?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -304,6 +420,29 @@ def unread(self, **generator_kwargs):
             self.subreddit._reddit, API_PATH['moderator_unread'].format(
                 subreddit=self.subreddit), **generator_kwargs)
 
+    def update_settings(self, **kwargs):
```
Can we just call this `update`? That should be consistent with the other models in PRAW4.

bboe replied:
Can we use `**settings`?

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -304,6 +420,29 @@ def unread(self, **generator_kwargs):
+    def update_settings(self, **kwargs):
+        """Update the subreddit's settings.
+
+        See :meth:`Subreddit._create_or_update` for details on each parameter.
```
As with `create`, can we explicitly list all the parameters in the docstring? They need not be explicitly listed in the function's signature, though.

bboe replied:
With this modification (https://github.com/praw-dev/praw/pull/642/commits/ca073d17707e4bec41504b6ac167184475f0913b#r72890863) it would be good to add a statement to the end of the docstring to say something to the effect of "additional keyword arguments can be passed to handle new settings as Reddit creates them".

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -17,6 +17,117 @@ class Subreddit(RedditBase, MessageableMixin, SubredditListingMixin):
     MESSAGE_PREFIX = '#'
 
+    @staticmethod
+    def _create_or_update(_reddit, allow_images=None, allow_top=None,
```
In the event that reddit adds new settings, can we permit this to take **other_settings? That way the client can manually add them in their call and things _should_ continue to work.

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -304,6 +420,29 @@ def unread(self, **generator_kwargs):
+        current_settings = self.settings()
+        fullname = current_settings.pop('subreddit_id')
+
+        # These attributes come out using different names than they go in.
```
This is unfortunate, but oh well.

bboe commented on praw/models/reddit/subreddit.py:
```diff
@@ -304,6 +420,29 @@ def unread(self, **generator_kwargs):
+        fullname = current_settings.pop('subreddit_id')
+
+        # These attributes come out using different names than they go in.
+        remap = {'lang': 'language',
```
Oh and can we sort these?

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -9,6 +11,45 @@
+            kwargs = {'subreddit_type': 'public',
+                      'wikimode': 'disabled'}
+
+            with pytest.raises(Redirect) as excinfo:
```
Do we need this request? If at the time we're testing create and the subreddit happens to exist, we can update the code to create one that doesn't exist.

bboe replied:
That being said, there should probably be a test for trying to create a subreddit that already exists.

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -9,6 +11,45 @@
+                # Redirects to /subreddits/search
+                next(self.reddit.subreddit(nonexistent_subreddit).new())
+
+            with pytest.raises(APIException) as excinfo:
```
Great test! Can we move this into it's own test, e.g., `test_create__already_exists`?

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -9,6 +11,45 @@
+                                             link_type='any', **kwargs)
+            assert excinfo.value.error_type == 'SUBREDDIT_EXISTS'
+
+            with pytest.raises(APIException) as excinfo:
```
Can we also move this into its own test?

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -9,6 +11,45 @@
+                                             link_type='any', **kwargs)
+            assert excinfo.value.error_type == 'NO_TEXT'
+
+            with pytest.raises(APIException) as excinfo:
```
Also in its own test?

bboe replied:
One reason for the separate tests is that I look through the cassettes when I add them to ensure there aren't any unexpected requests. I started reviewing the cassette first and was surprised to see redirects in it for this test. Then I realized what was going on :).

Overall, I really like the different test cases you have, and they will be perfect in their own test function.

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -229,14 +270,30 @@ def test_unignore_reports(self):
             assert count > 0
 
+    @mock.patch('time.sleep', return_value=None)
+    def test_update_settings(self, _):
```
Please update the test name when the function name is updated.

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -229,14 +270,30 @@ def test_unignore_reports(self):
+        with self.recorder.use_cassette(
+                'TestSubredditModeration.test_update_settings'):
+            before_settings = self.subreddit.mod.settings()
+            self.subreddit.mod.update_settings(title='A')
```
Should we assert that the title was not 'A' prior to attempting to update it?

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -229,14 +270,30 @@ def test_unignore_reports(self):
+            del after_settings['title']
+            assert before_settings == after_settings
+
+            self.subreddit.mod.update_settings(title='B')
```
While this resets the test for the next run, let's not worry about it since we only need it to be perfect once when we generate the cassette. It's easier to do that manually, than to try to make the test code reset the state for the future -- that got really annoying with the PRAW<4 tests.

bboe commented on tests/integration/models/reddit/test_subreddit.py:
```diff
@@ -229,14 +270,30 @@ def test_unignore_reports(self):
+            after_settings = self.subreddit.mod.settings()
+
+            # Ensure that nothing has changed besides what was specified.
+            del before_settings['title']
```
No need to change it but you could do `before_settings['title']  = 'A'` to save one line :)

bboe commented on tests/integration/models/reddit/test_multi.py:
```diff
@@ -9,7 +9,7 @@ class TestMultireddit(IntegrationTest):
     def test_add(self, _):
         self.reddit.read_only = False
-        with self.recorder.use_cassette('TestMulireddit.test_add'):
+        with self.recorder.use_cassette('TestMultireddit.test_add'):
```
Great catch on all those. Thanks!

bboe commented:
This is really awesome work! Great job. I finished putting my comments inline.

voussoir commented:
- Moved the big docstring to `SubredditModeration.update`
- Improved clarity of the docstring. 
- For `SubredditHelper.create`, only the required parameters are documented, user can read `update` for the rest.
- In the last commit, I was mistaken about the meaning of `show_media` which actually refers to thumbnails, not the new embedded expando. `show_media` now behaves correctly, and a new parameter `show_thumbnails` was added.
- Added `**other_settings` where necessary to give basic support for any new settings that may come along.
- Removed the "Redirect for visiting nonexistent subreddit" test since that isn't actually relevent to subreddit creation.
- Separated all of the creation error tests

bboe commented on praw/models/helpers.py:
```diff
@@ -94,16 +94,29 @@ def __call__(self, display_name):
+        :param title: The title of the subreddit.
+        :param wikimode: One of  ``anyone``, ``disabled``, ``modonly``.
+
+        See :meth:`~.SubredditModeration.update` for documenation on other
```
Typo `documentation`. Could we say, "See .. for documentation of other available settings.

bboe commented:
Great work!
